### DISCUSSION TITLE
Connecting to radio while debugging

### DISCUSSION THREAD

**Discussion by GhadeerElmkaiel** [bitcraze/discussions#1025] [Q&A] [answered]
### Main question:
Is it possible to connect the crazyflie to the radio while debugging it using an STLInk device?
### Problem I am facing:
I am currently developing costume functionalities for the crazyflie. When running the crazyflie, I get the heartbeat led sequence (Front right red led pulses with a frequency of 2 Hz). However, when I connect with the radio, the front right red led turns on continuously (Which is the low battery sequence). I am sure that it has nothing to do with the battery voltage.
I want to debug this issue with the ST-Link. However, I am not able to connect with the crazyflie via while I am debugging it. I have tried both, costume radio code, and cfclient interface, and both don't work.
Is it possible somehow to connect with the radio while debugging the code?

My setup for debugging:
ST-LInk V3, Crazyflie debugging kit, VS code software.

**Comment by krichardsson** [comment]
It should not be any problem to run the debugger while you are connected to the radio.
Have you got the debugger to work without the radio? There is [a guide in the documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/openocd_gdb_debugging/)

**Reply by GhadeerElmkaiel** [reply]
I followed the documentation to debug the code. 
I am able to debug the code via the VS code, However, when I try to connect with the crazyflie via radio I cannot connect.
Usually (When not debugging), when connecting to the crazyflie using the cfclient interface, the following three things happen:
1- The front left green/red leds turn on and flicker really fast 
2- The "**Connect**" button in the interface switches to say "**Disconnect**".
![Connect](https://github.com/bitcraze/discussions/assets/28564901/abc67e20-a94b-495f-9ecb-b43485a5272b)
![Disconnect](https://github.com/bitcraze/discussions/assets/28564901/a11684eb-27c9-4936-981d-18a3623e3e88)

3- The interface shows the parameters it reads from the crazyflie (estimated position and orientation)
In my case, when I am debugging the code, I get the heart beat led sequence which is a good indication. However, when I try to connect using the cfclient interface I have the following:
1- The radio led doesn't turn on (front left leds)
2- The "**Connect**" button in the interface switches to say "**Cancel**" which indicates that it is trying to connect, but it is not able to do that.
![Cancel](https://github.com/bitcraze/discussions/assets/28564901/60db62bc-b5a2-4b34-8415-f12a5f61a158)

3- The interface does not read any data from the crazyflie.

I tried to use an old official crazyflie code and still I couldn't connect to the radio while debugging.

**Reply by GhadeerElmkaiel** [reply]
I just noticed that the radio address that the interface finds is different when in the debugging mode, it is //0/0/2M instead of //0/80/2M.
I am not sure what does this mean.

**Reply by knmcguire** [reply]
Hi! This is the meaning of the address : radio:/number of crazyradio / channel / data-rate / crazyflie address. Not sure why it is showing 0 in your case but probably for now we can deam it unrelated to your initial issue and it might just be a cfclient bug.

Are you sure that you are not just hitting a breakpoint that 'freezes' the firmware cycles? If Freertos is not able to reach the communication part of the code than it makes sense that it won't communicate.

We will need to check this out and confirm the issue your having (and make a fixticket for it). In the meantime, you can try if you are able to connect through USB cable and communicate with the crazyflie through that?

**Reply by krichardsson** [reply]
`radio://0/0/2M` might indicate that your NRF is in bootloader mode, not sure why though.

I did some tests on my machine and I can not really reproduce your problem.

Just to double check, are you connecting the debugger to the STM debug port? I assume this is what you want to do, right?
The STM debug port is located under the USB connector on the main PCB.
The debug port that is located on a separate PCB that you mount on the side of the Crazyflie is for the NRF.

**Reply by GhadeerElmkaiel** [reply]
Hi @knmcguire. Thanks for the reply.
I am not sure that the issue is caused by a cfclient bug, as I also tried to connect to the radio with a costume python code and this didn't work.  My costume code works fine when I am not debugging.
It is also not a breakpoint issue, as I turned off all breakpoints and still, I was not able to connect. 
I will try to connect through USB cable and check if it is possible.

**Reply by GhadeerElmkaiel** [reply, marked as answer]
Hi @krichardsson, I think the issue was that I was putting the CF in the bootloader mode before starting the debugging process.
For some reason, at the beginning, the debugging didn't work when I turned on the crazyflie normally but only when I was in the bootloader mode, so I assumed that it is the only way to debug,
After reading your reply I understood that it is not the case.
I tried to run the debugging when the CF is turned on normally. It works now, and I can connect to the radio.
It was a wrong assumption on my side from the beginning that the CF should be in the bootloader mode in order for the debugging to work.
Thanks for your help. I think everything should be okay now.
